src/hotspot/share/opto/compile.cpp

```diff
@@ -4948,6 +4949,34 @@ bool Compile::coarsened_locks_consistent() {
   return true;
 }
 
+// Mark locking regions (identified by BoxLockNode) as coarsened
+// if locks coarsening optimization removed lock/unlock from them.
+// Such regions become unbalanced and we can't execute other
+// locks elimination optimization on them.
+void Compile::mark_coarsened_boxes() {
+  int count = coarsened_count();
+  for (int i = 0; i < count; i++) {
+    Node_List* locks_list = _coarsened_locks.at(i);
+    uint size = locks_list->size();
+    if (size > 0) {
+      AbstractLockNode* alock = locks_list->at(0)->as_AbstractLock();
+      BoxLockNode* box = alock->box_node()->as_BoxLock();
+      if (alock->is_coarsened() && !box->is_unbalanced()) { // Not marked already
```

Comment:
Is it possible that the position 0 `box` is already marked as unbalanced, but there are other `this_box` for positions `j=1..` that are not yet marked but should be?
Is it really ok that we are noch cheching all-with-all but only first-with-all?

Reply:
You are right. I removed `!box->is_unbalanced()` check.